internal/engine/engine.go

```diff
@@ -429,15 +436,70 @@ func untar(ctx context.Context, dst string, r io.Reader) error {
 
 			// if it's a link create it
 		case tar.TypeSymlink:
-			err := os.Symlink(header.Linkname, filepath.Join(dst, header.Name))
+			nobaseLinkname, nobaseName := resolveLinkPaths(header.Linkname, header.Name)
+			fullLinkname := filepath.Join(dst, nobaseLinkname)
+			fullName := filepath.Join(dst, nobaseName)
+			// Safeguard for cases where we're trying to link to something
+			// outside of our base fs.
+			if !strings.HasPrefix(fullLinkname, dst) {
+				logger.V(log.DBG).Info("Error processing symlink. Symlink would reach outside of the image archive. Skipping this link", "link", header.Name, "linkedTo", header.Linkname, "resolvedTo", fullLinkname)
+				continue
+			}
+			// Create the new link's directory if it doesn't exist.
+			dirname := filepath.Dir(fullName)
+			if _, err := os.Stat(dirname); err != nil {
+				if err := os.MkdirAll(dirname, 0o755); err != nil {
+					return err
+				}
+			}
+			err := os.Symlink(fullLinkname, fullName)
+			if err != nil {
+				logger.V(log.DBG).Info(fmt.Sprintf("Error creating symlink: %s. Ignoring.", header.Name), "link", fullName, "linkedTo", fullLinkname, "reason", err)
+				continue
+			}
+		case tar.TypeLink:
+			// We assume hard links will not contain relative pathing in the archive.
+			original := filepath.Join(dst, header.Linkname)
+			// Create the new link's directory if it doesn't exist.
+			if !strings.HasPrefix(original, dst) {
+				logger.V(log.DBG).Info("Error processing symlink. Symlink would reach outside of the image archive. Skipping this link", "link", header.Name, "linkedTo", header.Linkname, "resolvedTo", original)
```

Comment:
This log line is incorrect right? We want to say `hard link` like on L476? Or do we not need that clarity here?

Reply:
Fixed. If we operate with the assumption that all hard links are archived using the root of the tarball as the base path, then we actually don't need this logic here at all for hard links. I've removed it in the latest push.